receptor_affinity/mesh.py

```diff
@@ -122,18 +122,20 @@ def wait_for_ports(self) -> None:
             wanted_ports.add(self.stats_port)
         wait_for(lambda: wanted_ports.issubset(set(self.bound_ports())), num_sec=10)
 
-    def stop(self):
-        print(f"{time.time()} killing {self.name} ({self.uuid})")
+    def stop(self) -> None:
+        """Kill this node with SIGKILL.
+
+        If this node already appears to be dead, then do nothing.
+        """
         try:
             pgid = self.pgid
-        except ProcessLookupError:
-            print(f"Couldn't find a PGID for node {self.pid}; not stopping node.")
+        except NodeUnavailableError:
```

Comment:
the pgid property below does not raise that exception but the prior one, i suspect that was unintended

Reply:
At second glance it seems important ot mention that both excetions should be handled, `NodeUnavailableError` for idempotence, `ProcessLookupError` for a bad state

Reply:
If a process lacks a PID, then that indicates that the node has been shut down. But if a process has a PID and its PPID can't be found, then what does that mean? I only want to handle an exception if I know its meaning.

Reply:
a possible situation that makes this arise could be a process that had been cleaned up but is still in the mapping
is it possible that 2 threads may call stop on the same managed process?

o

Reply:
Oh neat, I hadn't thought about that. It's possible for two threads to call `Node.stop()` at the same time, but I doubt that that particular bit of parallelization will happen. I just don't see how doing that would allow one to write more useful and interesting test cases than is otherwise possible. But if I'm wrong, we can change this method. (: I'd tend to be conservative and only add in this particular bit of exception handling if we have a test case that actually needs it, but ultimately, I'm happy either way.

Reply:
i can go with that reasoning, if we are lucky yagni will sort it out, if not we can fix it